### ISSUE TITLE
Using AWSLimitChecker behind proxy

### ISSUE THREAD

**Issue by lastboy** [jantman/awslimitchecker#207]
Hi,
I have an issue when working behind a proxy.

I have declared environment variables like so:
export http_proxy="http://...
export https_proxy="http://..

This is the exception I get:
botocore.vendored.requests.exceptions.ConnectionError: HTTPSConnectionPool(host='email.ap-south-1.amazonaws.com', port=443): Max retries exceeded with url: / (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 503 Service Unavailable',)))

I bypass this issue by an ugly override like it described in here: https://github.com/jcgregorio/httplib2/issues/173

I tried to setup the environment programmatically, but still with no luck.

(Using Python 3.5.2)
Thank you,
-Arik

**Comment by jantman** [comment]
@lastboy I've never done anything AWS-like behind a proxy, and I'm not entirely sure how a proxy would work for HTTPS...

This is beyond the scope of what I really understand (It's connecting over HTTPS in the example you gave, so I don't see how a proxy could conceivably work). I'm sorry, but the best I can really do is point you to the underlying libraries this project uses to interact with AWS, boto3 (and botocore underlying that).

There are some issues relating to proxy if you search - https://github.com/boto/boto3/issues?utf8=%E2%9C%93&q=is%3Aissue%20proxy and https://github.com/boto/botocore/issues?utf8=%E2%9C%93&q=is%3Aissue%20proxy

botocore also has an open feature request entitled Proxy support - https://github.com/boto/botocore/issues/541 - which seems to indicate it works for some people?

**Comment by lastboy** [comment]
@jantman Thank you for the information, it is helpful 

I managed to resolve my issue - but I cannot tell exactly what did the trick. 
I think it was upgrading my modules

When I'll spot my fix I'll update it in here.

**Comment by jantman** [comment]
Ok, thanks! Would certainly be happy to have a solution here in case anyone else has the same problem. I'm going to close the issue, but please comment back if you figure out what solved the problem.

**Comment by lastboy** [comment]
Hi,

It turns out that it has nothing to do with the proxy - it is account related or something else I'm missing here.

When using awslimitchecker I can connect to the default region only, but to other regions I cannot - I get the 503 error.

I have a node client that works perfectly with using the access and secret key, since our user can connect to all regions.

I did a little boto3 test that also works:

import boto3
ec2 = boto3.resource('ec2', region_name='ap-southeast-2')
for instance in ec2.instances.all():
	print(instance, instance.tags)

Can you think of any direction what is the problem?

Thank you,
-Arik

**Comment by jantman** [comment]
Sorry for the delay.

Can you tell me how exactly you're running awslimitchecker? And how your account credentials are configured?

How do you mean the "default region"? Are you using the ``-r|--region`` parameter?

Also, have you tested your account / the same credentials **without** a proxy?

**Comment by jantman** [comment]
@lastboy 

So... I'm pretty sure I found the problem. I'm sorry for not realizing this earlier...

I took another look at the error that you happened to post:

```
botocore.vendored.requests.exceptions.ConnectionError: HTTPSConnectionPool(host='email.ap-south-1.amazonaws.com', port=443): Max retries exceeded with url: / (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 503 Service Unavailable',)))
```

So... that's throwing an error trying to connect to email.ap-south-1.amazonaws.com. [SES doesn't exist in that region](http://docs.aws.amazon.com/general/latest/gr/rande.html#ses_region). That also explains why your tests of connecting to EC2 succeeded. SES only exists in the US and EU regions.

There's already [code in the SES service class](https://github.com/jantman/awslimitchecker/commit/5f7024698fc924c84543ee58e51816d24f19cdf1) that handles this condition gracefully. However, it appears that having the proxy in the middle is hiding the actual error message from AWS - without the proxy this would result in botocore [detecting the problem](https://github.com/boto/botocore/blob/28b8831f5f8518bef5accf994fc377c58a3569f7/botocore/endpoint.py#L163) and throwing an [EndpointConnectionError](https://github.com/boto/botocore/blob/2def57f4f899ad749db8d750010c3712d854f703/botocore/exceptions.py#L63) (which awslimitchecker knows to ignore).

However, botocore is seeing a ConnectionError ``(Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 503 Service Unavailable',))``, which doesn't look like any error that it knows how to deal with, so botocore just raises it.

As far as I can tell, the problem here is that your proxy isn't returning the same response that AWS would. I'm not really sure what to suggest... aside from either fixing your proxy, or not calling awslimitchecker for services that don't exist in a given region.

**Comment by lastboy** [comment]
Hi

Reading your post. It looks like, if I will run my code w/o a proxy. I will be able to see the actual error.

As to your former question 
I have a default profile with our user credintial 
Below ~/.aws
That's about it.. And from the code im passing the region to the awslimitchecker 

Arik

**Comment by jantman** [comment]
Ok.

If you run the code without a proxy, you shouldn't get an error. awslimitchecker knows to ignore that error from AWS; the problem you're seeing is because your proxy is changing the error that AWS returns, so the code doesn't know how to handle it.

Your options, as far as I can see, are:
1. don't use a proxy.
2. have someone fix the proxy so it returns the same error that AWS does, instead of a different one.
3. don't run awslimitchecker for the services that don't exist in specific regions. This is a bit of a pain in a shell script, but could easily be done via the Python API.

-Jason

**Comment by lastboy** [comment]
Sorry for the late response.

I went over the botocore and boto3 code (just to understand the real cause) and I think it's not about the in specific proxy I'm using. It relates to any proxy usage.
I think that there are retries from the core that cause that kind of behavior (urllibs || requests).

I'm not sure that I found the spot , I'll get to that later on.

Look for "limit_exceeded" 
https://github.com/boto/botocore/blob/2010938e9250d6384bfe055e03b36e65e78eb070/botocore/data/_retry.json

Look for "max_attempts"
https://github.com/boto/botocore/blob/c3babf30df041ff064882f5779b96e1eeb1e4886/botocore/retryhandler.py
